File: Rust/src/bin/19_remove_nth_node_from_end_of_list.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}
// 
impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}
impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, mut n: i32) -> Option<Box<ListNode>> {
      let old_head=head.clone();
      let n_head=head.unwrap();
      let d_val=n_head;

      while n>0 {
        n_head=n_head.next.as_mut().unwrap();
        n-=1
      }

      old_head
    }
}

struct Solution {}

fn main() {
  let values = vec![1, 2, 3, 4, 5];

  let mut dummy = ListNode::new(0);
  let mut current = &mut dummy;

  for &val in &values {
    current.next = Some(Box::new(ListNode::new(val)));
    current = current.next.as_mut().unwrap();
  }

  let l = dummy.next;    
  println!("{:?}",l);

  Solution::remove_nth_from_end(l, 2);


}

File: Python/19.RemoveNthNodeFromEndofList.py
def pList(node):
    if node is None:
        print()
        return
    print(node.val, end='')
    pList(node.next)

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def removeNthFromEnd(self, head, n: int) -> ListNode:
        #edge case: list with 1 elem
        if not head.next and n==1:
            head=None
            return head

        last=head
        for i in range(n):
            last=last.next
        
        t_delete=head

        #edge case=delete last elem
        if not last and n==1:
            t_delete.next=None
            return head

        #print(t_delete)
        #print(head)
        # edge case=delete first elem
        if not last:
            head=head.next
            return head

        while (last.next):
            last=last.next
            t_delete=t_delete.next

        t_delete.next=t_delete.next.next
        return head

s=Solution()
l1=ListNode(1,ListNode(2,ListNode(3,ListNode(4,ListNode(5)))))
pList(s.removeNthFromEnd(l1,2))

l2=ListNode(1)
pList(s.removeNthFromEnd(l2,1))

l3=ListNode(1,ListNode(2))
pList(s.removeNthFromEnd(l3,1))

l4=ListNode(1,ListNode(2))
pList(s.removeNthFromEnd(l4,2)) # 2
